Refine and clarify operations in the async cache implementation

jxblum opened the pull request:
See #2743

jxblum committed:
Prepare issue branch for 2743.

jxblum committed:
Refine and clarify operations in asynchronous caching implementation.

Uses more descriptive names for operations, especially Reactive operations, by either calling a local, private method or introducing a (functional) variable with strongly typed parameters.

Edits and refines Javadoc.

Original Pull Request: #2717

Closes #2743

mp911de commented on src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java:
```diff
@@ -116,12 +103,12 @@ class DefaultRedisCacheWriter implements RedisCacheWriter {
-			asyncCacheWriter = UnsupportedAsyncCacheWriter.INSTANCE;
-		}
+	private boolean isAsyncCacheSupportEnabled() {
+		return this.connectionFactory instanceof ReactiveRedisConnectionFactory;
```
We're testing assignability against an interface that defines potentially absent types such as `Mono` and `Publisher`. Loading interfaces can work under certain circumstances, however depending on the runtime verifier mechanics, loading `ReactiveRedisConnectionFactory` without Project Reactor can cause `NoClassDefFoundError`.

Therefore, we used a safe scheme that is being used across the majority of portfolio projects and there's no reason to deviate from that.

jxblum replied:
Understood

jxblum replied:
Just thinking... perhaps the class presence check should be on a Reactive type (e.g. `Mono`) then, given it does depend on the runtime verifier, particularly in strange cases where it can load the `ReactiveConnectionFactory` interface, which has no explicit Reactive types until you drill into `ReactiveConnection`. The real question is, the `ReactiveConnection` as a return type on `ReactiveConnectionFactory` enough?

Also, by this same argument, then the only way a `ReactiveConnectionFactory` can exist is if _Lettuce_ is on the classpath anyway, which has a hard dependency on Reactor. Of course, there is no guarantee that Lettuce will always pull in Reactor as a transitive dependency (hence [this](https://github.com/spring-projects/spring-boot/blob/v3.1.4/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisReactiveAutoConfiguration.java#L43)), but also the reason I argued for [this](https://vmware.slack.com/archives/C055WU98P/p1694723156015419) (internal link).

jxblum replied:
I convinced it should be both `ReactiveConnectionFactory` and a Reactive type like `Mono` or `Flux`.

mp911de commented on src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java:
```diff
@@ -401,6 +382,14 @@ private static boolean shouldExpireWithin(@Nullable Duration ttl) {
 		return ttl != null && !ttl.isZero() && !ttl.isNegative();
 	}
 
+	private Expiration toExpiration(Duration ttl) {
```
These tiny methods are unnecessary and generate a lot of noise making it harder to read the code.

jxblum replied:
Technically, I have even seen things likes this:

```java
return connection.stringCommands().set(ByteBuffer.wrap(key), ByteBuffer.wrap(value),
         Expiration.from(ttl.toMillis(), TimeUnit.MILLISECONDS), SetOption.upsert());
```

jxblum replied:
`toExpiration(..)` is a easy to read method, `private`, and at the end of the class.

jxblum replied:
Additionally, these methods can be use as _Lambda's_, such as in `.map(:Function<S, T>)` functions, which are useful in the context of Optionals, Streams, etc.

mp911de reviewed (changes requested):
Did a first review pass.

jxblum commented on src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java:
```diff
@@ -545,28 +539,52 @@ private Mono<Void> doUnlock(String name, ReactiveRedisConnection connection) {
 					.then();
 		}
 
+		private Flux<Long> waitInterval(Duration period) {
+			return Flux.interval(Duration.ZERO, period);
+		}
+
+		private ByteBuffer toCacheLockKey(String cacheName) {
+			return ByteBuffer.wrap(createCacheLockKey(cacheName));
+		}
+
+		private ReactiveRedisConnectionFactory getReactiveConnectionFactory() {
+			return (ReactiveRedisConnectionFactory) DefaultRedisCacheWriter.this.connectionFactory;
+		}
+
+		private Mono<ReactiveRedisConnection> getReactiveConnection() {
+			return Mono.fromSupplier(getReactiveConnectionFactory()::getReactiveConnection);
+		}
+
```
I disagree. They help reduce the noise in the method where this code is clumped up, thereby causing the code formatter to break lines at ridiculous points. Additionally, using a named method is much clearer than some arbitrary variable, like `cf`.

jxblum replied:
In these methods... 

With `waitForInterval(period)`, it is more descriptive than `Flux.interval(Duration.ZERO, period)`. What is `Duration.ZERO` anyway? Oh right, delay. Why do I even need to know it is `Flux` for that matter? It is an implementation detail. You don't know what `Object.wait(..)` does when you call it.

`getReactiveConnectionFactory()` avoid the cast could serve as an extension point if `protected`. In too many places, we are referring to state variables directly rather than the appropriate getter (where applicable), preventing extensibility.

`getReactiveConnection()` should go without saying, `Mono` into a `connectionFactory.getConnection()`. Having this logic spread all over the class is redundant and busy creating noise in that need wrapped connections.

What does it matter whether it is a variable or a method. Methods help remove clutter from other methods and can be reused.

jxblum replied:
What was there [before](https://github.com/spring-projects/spring-data-redis/pull/2744/files#diff-6c6aae73dbe5cee3bc00315ecfb74e08fb1606bc25c1abe443b2c14afea8e6c0L564-R587) was certainly less readable.

jxblum commented on src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java:
```diff
@@ -74,33 +73,21 @@ class DefaultRedisCacheWriter implements RedisCacheWriter {
-	 * @param cacheStatisticsCollector must not be {@literal null}.
-	 * @param batchStrategy must not be {@literal null}.
+	 * @param sleepTime sleep time between lock request attempts. Must not be {@literal null}.
+	 * Use {@link Duration#ZERO} to disable locking.
```
The Spring Data source code formatter for Javadoc needs to be fixed.

The excessive indentation is causing unnatural line breaks (such as sentences ending in `and`), creating orphans, and doing other inconsistent things that do not align with the core Spring Frameworks Javadoc guidelines.

jxblum commented on src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java:
```diff
@@ -74,33 +73,21 @@ class DefaultRedisCacheWriter implements RedisCacheWriter {
 	}
 
 	/**
-	 * @param connectionFactory must not be {@literal null}.
```
"must not be null" on `@param` tags is redundant and non-informative.

In general, I'd argue if we are not going to take the time to write meaningful Javadoc, then there is no point to have Javadoc on the method at all.

jxblum commented:
> Did a first review pass.

Thank you for the feedback. I reviewed each comment and cited my reasoning for the requested changes.

I reworked this PR to be more inline with our expectations.

Specifically, I:

1) Added a more robust check to the detection of Reactive types and usage, not unlike Spring Boot ([here](https://github.com/spring-projects/spring-boot/blob/v3.1.4/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisReactiveAutoConfiguration.java#L43)).

2) I reduced the number of `private` (helper) methods only slightly and reorganized them based on where they belong. Certain `private` methods remain because 1) they are reused 2) and/or are more descriptive (especially in the Reactive context).

3) Please! Can we follow the Spring Framework [Javadoc guidelines](https://github.com/spring-projects/spring-framework/wiki/Code-Style#javadoc).

I updated this PR.
